crowbar_framework/app/models/api/crowbar.rb

```diff
@@ -183,6 +183,37 @@ def openstack_check
           disks += n["swift"]["devs"].size
         end
         ret[:too_many_replicas] = replicas if replicas > disks
+      end
+
+      def mixed_roles_check
+        conflicting_roles = [
+          "cinder-controller",
+          "glance-server",
+          "keystone-server",
+          "neutron-server",
+          "neutron-network",
+          "nova-controller",
+          "swift-proxy",
+          "swift-ring-compute",
+          # FIXME: maybe we might ignore the confict for these roles:
+          "ceilometer-server",
+          "heat-server",
+          "horizon-server",
+          "manila-server",
+          "trove-server",
+          # FIXME: use this as better check to make sure compute node is not part of cluster?
+          "pacemaker-cluster-member",
+        ]
+        ret = {}
+        ["kvm", "xen"].each do |virt|
+          NodeObject.find("roles:nova-compute-#{virt}").each do |node|
+            conflict = node.roles & conflicting_roles
+            unless conflict.empty?
+              ret[:role_conflicts] ||= {}
+              ret[:role_conflicts][node.name] = conflict
```

Comment:
So far nowhere, that still missing. Adding that part is easy, first I wanted to discuss this part (those FIXME's in this PR)